src/MongoDB.Driver.Core/Core/Servers/RoundTripTimeMonitor.cs

```diff
@@ -0,0 +1,151 @@
+﻿/* Copyright 2020-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Diagnostics;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using MongoDB.Driver.Core.Connections;
+using MongoDB.Driver.Core.Misc;
+
+namespace MongoDB.Driver.Core.Servers
+{
+    internal interface IRoundTripTimeMonitor : IDisposable
+    {
+        TimeSpan Average { get; }
+        void AddSample(TimeSpan roundTripTime);
+        void Reset();
+        Task Run();
+    }
+
+    internal class RoundTripTimeMonitor : IRoundTripTimeMonitor
```

Comment:
the spec PR regarding this logic: https://github.com/mongodb/specifications/pull/807